### PR TITLE
Lodestar Grail should work.

### PR DESCRIPTION
**Pull request by ShubusCorporation**
When Lodestar Grail is built:

- Any terrain is native for all creatures of 'Cove' faction. Thus, any hero with only 'Cove' creatures has not terrain movement penalty.
- Any creature of 'Cove' faction receives +1 attack, +1 defence, +1 speed bonuses in the battle.

When Nomad creature is in the hero army: No terrain penalty when hero is on the sand (according to the FizMiG)

### CODE REVIEW

File: lib/GameConstants.h [hunk 1]
```diff
@@ -861,7 +861,7 @@ class DLL_LINKAGE ETerrainType
 	enum EETerrainType
 	{
 		WRONG = -2, BORDER = -1, DIRT, SAND, GRASS, SNOW, SWAMP,
-		ROUGH, SUBTERRANEAN, LAVA, WATER, ROCK
+		ROUGH, SUBTERRANEAN, LAVA, WATER, ROCK, ANY_TERRAIN
```
Comment (alexvins) [inline lib/GameConstants.h:864]: ANY_TERRAIN should have negative value like WRONG & BORDER

File: lib/CCreatureHandler.h [hunk 2]
```diff
@@ -34,6 +34,7 @@ class DLL_LINKAGE CCreature : public CBonusSystemNode
 
 	CreatureID idNumber;
 	TFaction faction;
+	mutable ETerrainType::EETerrainType nativeTerrain;
```
Comment (alexvins) [inline lib/CCreatureHandler.h:37]: this field need serialization, and to prevent desync add additional mutable fileld

File: lib/mapObjects/CGHeroInstance.cpp [hunk 3]
```diff
@@ -89,43 +89,62 @@ ui32 CGHeroInstance::getTileCost(const TerrainTile & dest, const TerrainTile & f
 			break;
 		}
 	}
-	else if(ti->nativeTerrain != from.terType && !ti->hasBonusOfType(Bonus::NO_TERRAIN_PENALTY, from.terType))
+	else
 	{
-		static const CSelector selectorPATHFINDING = Selector::typeSubtype(Bonus::SECONDARY_SKILL_PREMY, SecondarySkill::PATHFINDING);
-		static const std::string keyPATHFINDING = "type_"+std::to_string((si32)Bonus::SECONDARY_SKILL_PREMY)+"s_"+std::to_string((si32)SecondarySkill::PATHFINDING);
+		bool nomadBonus = from.terType == ETerrainType::SAND && ti->heroHasCreature(CreatureID::NOMAD);
```
Comment (alexvins) [inline lib/mapObjects/CGHeroInstance.cpp:94]: Why is this needed? Nomad already has NO_TERRAIN_PENALTY for sand.

File: lib/mapObjects/CGTownInstance.cpp [hunk 4]
```diff
@@ -1232,6 +1232,16 @@ void CGTownInstance::recreateBuildingsBonuses()
 		addBonusIfBuilt(BuildingID::GRAIL, Bonus::PRIMARY_SKILL, +10, PrimarySkill::ATTACK); //grail
 		addBonusIfBuilt(BuildingID::GRAIL, Bonus::PRIMARY_SKILL, +10, PrimarySkill::DEFENSE); //grail
 	}
+	else if(this->town->getFactionName() == "Cove" && hasBuilt(BuildingID::GRAIL))
```
Comment (alexvins) [inline lib/mapObjects/CGTownInstance.cpp:1235]: `this->town->getFactionName() == "Cove"` such hardcoding should be avoided.

**Review by alexvins — CHANGES_REQUESTED** [review]

File: lib/mapObjects/CGTownInstance.cpp (hunk 4 above)
Comment (ShubusCorporation) [reply lib/mapObjects/CGTownInstance.cpp:1235]: Will be changed later, with general solution for all grails. There's no better solution now. Cove Grail should work because it's very popular and semi-official town.

**Review by ShubusCorporation — COMMENTED** [review]
Faced a bug, so the solution was reworked

File: lib/CCreatureHandler.h (hunk 2 above)
Comment (ShubusCorporation) [reply lib/CCreatureHandler.h]: This field is removed, there's another field in the Stack class

**Comment by ShubusCorporation** [comment]
@alexvins

File: lib/mapObjects/CGTownInstance.cpp (hunk 4 above)
Comment (alexvins) [reply lib/mapObjects/CGTownInstance.cpp:1235]: You could use identifier, but do not use localized name.

Comment (ShubusCorporation) [reply lib/mapObjects/CGTownInstance.cpp]: Accepted.

**Review by alexvins — APPROVED** [review]
